include/bluetooth/services/latency.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @file
+ * @defgroup bt_gatt_latency BLE GATT Latency Service API
+ * @{
+ * @brief API for the BLE GATT Latency Service.
+ */
+
+#ifndef BT_GATT_LATENCY_H_
+#define BT_GATT_LATENCY_H_
+
+#include <bluetooth/uuid.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt_dm.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief Latency callback structure. */
+struct bt_gatt_latency_cb {
+	/** @brief Latency received callback.
+	 *
+	 * This function is called when a GATT write request has been received
+	 * by the Latency characteristic.
+	 *
+	 * @param[in] buf Latency data.
+	 */
+	void (*latency_request)(const void *buf, u16_t len);
+
+	/** @brief Latency received callback.
+	 *
+	 * This function is called when a GATT write response has been received
+	 * by the Latency characteristic.
+	 *
+	 * @param[in] buf Latency data.
+	 */
+	void (*latency_response)(const void *buf, u16_t len);
+};
+
+/** @brief Latency structure. */
+struct bt_gatt_latency {
+	/** Characteristic handle. */
+	u16_t handle;
+
+	/** Latency parameter. */
+	struct bt_gatt_write_params latency_params;
+
+	/** Connection object. */
+	struct bt_conn *conn;
+
+	/** Internal state. */
+	atomic_t state;
+};
+
+/** @brief Latency Service UUID. */
+#define LATENCY_UUID_DATA                                           \
+	((u8_t []) {0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+		    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67})
+#define BT_UUID_LATENCY                                                     \
+	BT_UUID_DECLARE_128(0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
+			    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67)
```

Comment:
Yup, I had the same idea at the beginning, but it doesn't get approved by the coding-style of Zephyr.

-:9: WARNING:COMPLEX_MACRO: Macros with complex values should be enclosed in parentheses
#9: FILE: include/bluetooth/services/latency.h:60:
+#define LATENCY_UUID_TEST 0x23, 0xD1, 0xBC, 0xEA, 0x5F, 0x78, 0x23, 0x15, \
+                         0xDE, 0xEF, 0x12, 0x12, 0x23, 0x15, 0x00, 0x00

- total: 0 errors, 1 warnings, 9 lines checked

Reply:
Fortunately, warnings are not block a merge. Sometimes, for greater code readability, it is worth ignoring. I think in this place you can simplify this defines to one.

Reply:
And more over the latency service will be consistent with our other custom service. You can check the lbs and the uart service.

Reply:
Being consistent with the rest of the services is a good idea. The warning can be ignored.

Reply:
I managed to resolve the coding-stype problem with a magic macro in util.h as below. It would look better now. :)

#define LATENCY_UUID                                                \
	UTIL_EXPAND(0x3A, 0x81, 0xC0, 0x58, 0xDE, 0xFD, 0x46, 0x34, \
		    0x9B, 0xF3, 0xDB, 0x58, 0x01, 0x6E, 0x13, 0x67)

#define BT_UUID_LATENCY BT_UUID_DECLARE_128(LATENCY_UUID)

It can also be used in the example (main.c).

static const struct bt_data ad[] = {
	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
	BT_DATA_BYTES(BT_DATA_UUID128_ALL, LATENCY_UUID),
};